### PR TITLE
Add `dirtyRev` and `dirtyShortRev` to `fetchGit`

### LINKED ISSUE #4682: Should a flake's `self.rev` return something useful in modified repos?
What's the thinking on why a flake's `self.rev` won't evaluate in a dirty repo? IMO, it would be a lot nicer to work with `self.rev` in Nix expressions if it returned `<rev>-dirty` (or similar) in this situation.

I do understand the argument that it's better to fail in this situation than to return `<rev>-dirty`, but then it seems a bit inconsistent that you *can* get a modified repo's last-modified via `self.lastModified` or `self.lastModifiedDate`.

### PR DESCRIPTION
**Pull request by Enzime**
This also adds it to `nix flake metadata --json` as `.dirtyRevision` and when running `nix flake metadata` it'll output the dirty revision as the replacement for `Revision: `

Fixes #4682

### CODE REVIEW

**Comment by maydayv7** [comment]
(Bumping...)

**Comment by Enzime** [comment]
@maydayv7 if you're interested in using this PR, take a look at the way I'm using it from my dotfiles:

I use an overlay to patch Nix in `nixpkgs`: https://github.com/Enzime/dotfiles-nix/blob/7b0b6befe0bf864c83ec03c913fc06128942483d/overlays/nix/flake.nix#L19-L25

And here is where I actually use `dirtyRev`: https://github.com/Enzime/dotfiles-nix/blob/7b0b6befe0bf864c83ec03c913fc06128942483d/flake.nix#L51-L54

**Comment by maydayv7** [comment]
Thanks a lot!

**Comment by lilyball** [comment]
I really want this.

@Enzime Just as a tip, your second example of where you use `dirtyRev` can be simplified like `full = self.rev or self.dirtyRev or "dirty-inputs";`.

**Comment by lilyball** [comment]
It looks like this always picks 7 digits for the short revision? Git defaults to 7 but will increase the length depending on the number of objects in the repository to avoid collisions in large repos. If we don't want to run a second `git describe` command to let git pick the abbreviation, then we should probably pick a longer default (I tend to go for 9 when I'm shortening revs, but I don't know what logic git uses to pick this number).

For reference, in current github:NixOS/nix git picks 9 instead of 7 for abbreviations, and in github:NixOS/nixpkgs it picks 11. So it's probably best to actually let git go ahead and run its logic here.

File: src/libfetchers/git.cc
Comment (lilyball) [inline src/libfetchers/git.cc:250]: This command is surprisingly slow in github:NixOS/nixpkgs, I'm really not sure why given the `--match=''` means it always ends up outputting the full sha, and it's not the `--dirty` flag that makes it slow (it's slow without that too).

I'm inclined to say this should instead run `git rev-parse --verify HEAD` for the sha, and then some other command to compute dirty status (perhaps `git diff --quiet HEAD`).

I'm also not sure how submodules affect this. AIUI current behavior is to ignore submodules unless the flake is specified with `?submodules=1`, but I don't know if that affects the current revision calculation or if that propagates to this command somehow.

Comment (Enzime) [reply src/libfetchers/git.cc]: I've chosen to switch this to use `git rev-parse --verify HEAD` as suggested :+1: 

As this code only runs if the worktree is dirty so we don't need to check that it's dirty and we can manually append "-dirty" unconditionally

**Comment by Enzime** [comment]
@lilyball I've also fixed the short revisions not matching Git by adding an extra call to `git rev-parse` with `--short` :+1:

**Comment by Enzime** [comment]
@edolstra @thufschmitt any chance I could get a review on this PR? the linked issue #4682 shows there is interest in being able to access the revision even in dirty worktrees

**Comment by fricklerhandwerk** [comment]
Triaged in the Nix team meeting (notes for maintainers):

- @edolstra: have to be careful about adding metadata, potentially expensive to maintain
    - probably not a problem here
    - @thufschmitt: is it an issue with `libgit`?
    - @edolstra: no, but should check to make sure

File: src/libexpr/primops/fetchTree.cc
Comment (thufschmitt) [inline src/libexpr/primops/fetchTree.cc:60]: I'm not sure what that comment means: `dirtyRev` should only exist when fetching a dirty local repo, right?

**Review by thufschmitt — APPROVED** [review]
Sorry for the delay. This looks good to me, except for one comment which seems a bit off.

This is missing documentation, but none of this is documented any way, so it will have to wait for https://github.com/NixOS/nix/pull/8509/ (or equivalent) to land first.

File: src/libexpr/primops/fetchTree.cc
Comment (Enzime) [reply src/libexpr/primops/fetchTree.cc:60]: Yeah, I originally added this comment just to help me when writing this PR, I'm happy to remove it if you think it'll be less confusing

Comment (thufschmitt) [reply src/libexpr/primops/fetchTree.cc:60]: Yeah, I think it would be better to drop it. Just ping me once it's done and I can merge

Comment (Enzime) [reply src/libexpr/primops/fetchTree.cc]: @thufschmitt updated :+1:

Comment (thufschmitt) [reply src/libexpr/primops/fetchTree.cc]: Thanks :)

**Comment by AleXoundOS** [comment]
Is it expected that neither `dirtyRev` nor `rev` exist when `--override-input` is used with `nix build`?

**Comment by AleXoundOS** [comment]
@Enzime, was the original intention to ignore the `--override-input` case in this PR? And how do you think it should be handled?

**Comment by con-f-use** [comment]
Hmm... `self.rev or self.dirtyRev or "dirty"` now gives me a mysterious `Failed assertions: system.copySystemConfiguration is not supported with flakes`

**Pull request merged**
